### PR TITLE
[trivial] Improve error message when aggregate template is used as a type

### PR DESCRIPTION
**Pull request by ntrel**
Make it clear with aggregate templates that they are templates and haven't been instantiated.

Note: `s.kind` is e.g. just "struct" for aggregate templates. Non-aggregates say "template".
Note: When `td.onemember` is an alias, we don't know if it resolves to a type or not until the template is instantiated.

http://forum.dlang.org/post/p49c55$m65$1@digitalmars.com

### CODE REVIEW

**Review by quickfur — COMMENTED** [review]
LGTM.

**Review by jacob-carlborg — APPROVED** [review]

**Comment by wilzbach** [comment]
CircleCi failure is unrelated - https://github.com/dlang/dmd/pull/7770

File: src/dmd/typesem.d [hunk 1]
```diff
@@ -1156,7 +1156,12 @@ private extern (C++) final class TypeSemanticVisitor : Visitor
         {
             if (s)
             {
-                s.error(loc, "is used as a type");
+                auto td = s.isTemplateDeclaration;
+                if (td && td.onemember && td.onemember.isAggregateDeclaration)
+                    mtype.error(loc, "%s template `%s` is used as a type without instantiation",
```
Comment (WalterBright) [inline src/dmd/typesem.d]: I suggest an improvement:
```
template %s `%s` is used as a type without instantiation; to instantiate it use `%s!(arguments)`
```

**Comment by wilzbach** [comment]
> CircleCi failure is unrelated - #7770 

In lack of someone wanting merging #7770, I rebased the PR. This will fix the CircleCi failure too.

**Review by WalterBright — CHANGES_REQUESTED** [review]
Please make the change I suggested. A good error message should say what is wrong, and suggest corrective action.

File: src/dmd/typesem.d (hunk 1 above)
Comment (MetaLang) [reply src/dmd/typesem.d]: What if the user has a variable called `arguments` already in scope? They might get confused.

**Review by WalterBright — APPROVED** [review]

File: src/dmd/typesem.d (hunk 1 above)
Comment (ntrel) [reply src/dmd/typesem.d]: @MetaLang Maybe *use `%s!()`* instead? That way if the template requires arguments they will get an error to that effect, and this error may come up when the template doesn't need arguments anyway, as in the forum example.

**Pull request merged**